### ISSUE TITLE
text similarity

### ISSUE THREAD

**Issue by divastar** [huggingface/setfit#198]
Hi can I use this system to obtain the similaarity scores of my data set to a given prompts.
If not what best solution could help this problem?
thank you

**Comment by fralik** [comment]
I am not an author of this library. However, I would share my thoughts.

This library is to fine-tune models (i.e. train models), which can then later used to obtain similarity scores. If you only have a data set and given prompts, then have a look at [sentence-transformers](https://www.sbert.net/docs/pretrained_models.html) library. Use one of the pre-trained models directly. Calculate embeddings of your text data and prompts, then you will be able to calculate similarity between embeddings (e.g. cosine similarity).

**Comment by divastar** [comment]
Hi Fralik,
thank you for your responce.
I do want to better the model for my usecase in my language
I do have data matching specific categories enough data to train for each category.
But at the end I will only need to specify similarity scores of any user's prompt to all my data categories
So I understand that I can finetune some multilingual model for example 'paraphrase-multilingual-MiniLM-L12-v2'
for my data categories with setfeat
and then use the [sentence-transformers](https://www.sbert.net/docs/pretrained_models.html) library
for the similarity. am I right?
Thank you

**Comment by fralik** [comment]
Yes, sounds right to me. In SetFit's readme file you are presented with multiple ways of fine-tuning. I think that you need to train using `SetFitHead`. This is an example where they show that body is fine-tuned with the head. Once you are done with fine-tuning, you can throw away the head and only use the underlying body to get embeddings.

**Comment by divastar** [comment]
thank you! I will work on that

**Comment by divastar** [comment]
Hi [fralik](https://github.com/fralik), thank you for your response. So my use case is to only get the similarity of sentences without labels. what I have is very large datasets of similar and non similar pairs of sentences. So how I should construct this dataset to input to the setfit trainer?
thank you.

**Comment by tomaarsen** [comment]
@divastar Hello,

I want to briefly elaborate on this issue. SetFit consists of two parts:
1. A [Sentence Transformer](https://sbert.net) body
2. A classification head

The ST body produces embeddings for sample sentences, and then the classification head chooses a corresponding label. If you don't have labels that you care about, then you won't care much about the classification head. What you really want, is to train a sentence transformer. This can be done via SetFit, but it would be very hacky. My recommended way is to follow the documentation for training a sentence transformer: https://sbert.net/docs/training/overview.html

This documentation shows that you may create training examples like so:
```python
from sentence_transformers import SentenceTransformer, InputExample
from torch.utils.data import DataLoader

train_examples = [
    InputExample(texts=['My first sentence', 'My second sentence'], label=0.8),
    InputExample(texts=['Another pair', 'Unrelated sentence'], label=0.3)
]
train_dataloader = DataLoader(train_examples, shuffle=True, batch_size=16)
```
The `train_examples` is a list of `InputExample` instances, which contain pairs of sentences. The label is then a float between 0.0 and 1.0 indicating the similarity of the two sentences. So, for your dataset of similar pairs you can choose e.g. 1.0, and 0.0 for non-similar pairs. 

A full (although minimal) training example is this:
```python
from sentence_transformers import SentenceTransformer, InputExample, losses
from torch.utils.data import DataLoader

#Define the model. Either from scratch of by loading a pre-trained model
model = SentenceTransformer('distilbert-base-nli-mean-tokens')

#Define your train examples. You need more than just two examples...
train_examples = [InputExample(texts=['My first sentence', 'My second sentence'], label=0.8),
    InputExample(texts=['Another pair', 'Unrelated sentence'], label=0.3)]

#Define your train dataset, the dataloader and the train loss
train_dataloader = DataLoader(train_examples, shuffle=True, batch_size=16)
train_loss = losses.CosineSimilarityLoss(model)

#Tune the model
model.fit(train_objectives=[(train_dataloader, train_loss)], epochs=1, warmup_steps=100)
```
(As copied from the [SentenceTransformer documentation](https://sbert.net/docs/training/overview.html))

Once the model has been fitted, you can follow the step for [Semantic Textual Similarity](https://sbert.net/docs/usage/semantic_textual_similarity.html), which involves getting embeddings of two sentences and then using cosine similarity to determine how similar the sentences are.

Hope this helps a bit.

- Tom Aarsen

**Comment by divastar** [comment]
Hello [tomaarsen](https://github.com/tomaarsen),
thank you very much for your response!
I will definitely follow on this.
But one more question: actually I need to compare pairs
of a kind of documents of varying length, which sometimes consist of several sentences
as well as some tabular data that I convert to phrases and add to these documents.
So will this setup still work as well?
Is this model is still the best choice for such a problem?

Thank you

**Comment by tomaarsen** [comment]
I must say that I'm not certain whether the approach works well with multi-sentence texts. However, I am quite certain that the tokenizer truncates longer inputs to a specific length, usually 512 word pieces (around 300-400 words), but sometimes also 128 word pieces. See some documentation [here](https://www.sbert.net/examples/applications/computing-embeddings/README.html#input-sequence-length).
